drivers/dai/intel/ssp/ssp.h

```diff
@@ -0,0 +1,384 @@
+
+#define SSTSA_TSEN			BIT(8)
+#define SSRSA_RSEN			BIT(8)
+
+#define SSCR3_TFL_MASK	MASK(5, 0)
+#define SSCR3_RFL_MASK	MASK(13, 8)
+#define SSCR3_TFL_VAL(scr3_val)	(((scr3_val) >> 0) & MASK(5, 0))
+#define SSCR3_RFL_VAL(scr3_val)	(((scr3_val) >> 8) & MASK(5, 0))
+#define SSCR3_TX(x)	SET_BITS(21, 16, (x) - 1)
+#define SSCR3_RX(x)	SET_BITS(29, 24, (x) - 1)
+
+#define SSIOC_TXDPDEB	BIT(1)
+#define SSIOC_SFCR	BIT(4)
+#define SSIOC_SCOE	BIT(5)
+
+/* For 8000 Hz rate one sample is transmitted within 125us */
+#define SSP_MAX_SEND_TIME_PER_SAMPLE 125
+
+/* SSP flush retry counts maximum */
+#define SSP_RX_FLUSH_RETRY_MAX	16
+
+#define SSP_CLK_MCLK_ES_REQ	BIT(0)
+#define SSP_CLK_MCLK_ACTIVE	BIT(1)
+#define SSP_CLK_BCLK_ES_REQ	BIT(2)
+#define SSP_CLK_BCLK_ACTIVE	BIT(3)
+
+#define I2SLCTL			0x71C04
+#define I2SLCTL_SPA(x)		BIT(0 + x)
+#define I2SLCTL_CPA(x)		BIT(8 + x)
+
+/** \brief Offset of MCLK Divider Control Register. */
+#define MN_MDIVCTRL 0x0
+
+/** \brief Enables the output of MCLK Divider. */
+#define MN_MDIVCTRL_M_DIV_ENABLE(x) BIT(x)
+
+/** \brief Offset of MCLK Divider x Ratio Register. */
+#define MN_MDIVR(x) (0x80 + (x) * 0x4)
+
+/** \brief Bits for setting MCLK source clock. */
+#define MCDSS(x)	SET_BITS(17, 16, x)
+
+/** \brief Offset of BCLK x M/N Divider M Value Register. */
+#define MN_MDIV_M_VAL(x) (0x100 + (x) * 0x8 + 0x0)
+
+/** \brief Offset of BCLK x M/N Divider N Value Register. */
+#define MN_MDIV_N_VAL(x) (0x100 + (x) * 0x8 + 0x4)
+
+/** \brief Bits for setting M/N source clock. */
+#define MNDSS(x)	SET_BITS(21, 20, x)
+
+/** \brief Mask for clearing mclk and bclk source in MN_MDIVCTRL */
+#define MN_SOURCE_CLKS_MASK 0x3
+
+#if CONFIG_INTEL_MN
+/** \brief BCLKs can be driven by multiple sources - M/N or XTAL directly.
+ *	   Even in the case of M/N, the actual clock source can be XTAL,
+ *	   Audio cardinal clock (24.576) or 96 MHz PLL.
+ *	   The MN block is not really the source of clocks, but rather
+ *	   an intermediate component.
+ *	   Input for source is shared by all outputs coming from that source
+ *	   and once it's in use, it can be adjusted only with dividers.
+ *	   In order to change input, the source should not be in use, that's why
+ *	   it's necessary to keep track of BCLKs sources to know when it's safe
+ *	   to change shared input clock.
+ */
+enum bclk_source {
+	MN_BCLK_SOURCE_NONE = 0, /**< port is not using any clock */
+	MN_BCLK_SOURCE_MN, /**< port is using clock driven by M/N */
+	MN_BCLK_SOURCE_XTAL, /**< port is using XTAL directly */
+};
+#endif
+
+struct mn {
+	uint32_t base;
+	/**< keep track of which MCLKs are in use to know when it's safe to
+	 * change shared clock
+	 */
+	int mclk_sources_ref[DAI_NUM_SSP_MCLK];
+	int mclk_rate[DAI_NUM_SSP_MCLK];
+	int mclk_source_clock;
+
+#if CONFIG_INTEL_MN
+	enum bclk_source bclk_sources[(DAI_NUM_SSP_BASE + DAI_NUM_SSP_EXT)];
+	int bclk_source_mn_clock;
+#endif
+
+	struct k_spinlock lock; /**< lock mechanism */
+};
+
+struct freq_table {
```

Comment:
This is also quite a common name we would need to namespace with a common prefix here.

Reply:
ok